### ISSUE TITLE
RuntimeException: only ACLs can be stored in the ACL key

### ISSUE THREAD

**Issue by Ralphilius** [ParsePlatform/Parse-SDK-Android#499]
Hello,

I got this error message when trying to save a new object into Parse Server. My code is as below:

    ParseGeoPoint geoPoint = new ParseGeoPoint(lat,long);
    newComment = new Comment();
    newComment.put("comment",message);
    newComment.put("state",0);
    newComment.put("rating",0);
    newComment.put("usersUpVoted",new ArrayList<String>());
    newComment.put("usersDownVoted",new ArrayList<String>());
    newComment.put("userImageName",userImage);
    newComment.put("post_id", ParseObject.createWithoutData(Post.class,postId));
    newComment.put("user_id",ParseUser.getCurrentUser());
    newComment.put("gps",geoPoint);
    newComment.saveInBackground(new SaveCallback() {
        @Override
        public void done(ParseException e) {
             // Processing with success
        }
    });

If I uninstalled then reinstalled the apk, it is working. Also, happening in production.

**Comment by Sumsar** [comment]
I have the same issue. When I update my application some save operation with nested objects fail.

**Comment by eddieux** [comment]
I'm seeing the same exception but with ParseObject.saveAllInBackground method. First observations: exception happens after rerunning my app in Android Studio with an user (of my app) still logged in. Exception does not happen after first install of the app. Just like  @Ralphilius mentioned, reinstalling the app cures the problem. Logging the user out and in again seems also working.

**Comment by sreejithraman** [comment]
I am seeing this bug even after the first install of the app, reinstalling didn't fix the issue. And this is happening live

**Comment by matrixdevz** [comment]
I am also facing this issue from last two days when call ParseObject.saveAllInBackground method.
Problem is same Just like @eddieux mentioned

**Comment by Aurintas** [comment]
Same here when using ParseObject.saveAllInBackground :/ What I've noticed is that I get this error whenever I try to save **newly** created objects (not the queried ones). And also, I don't get this error when saving same objects with saveInBackground one by one.

**Comment by DrBeak1** [comment]
Also seeing this issue.

**Comment by DrBeak1** [comment]
I haven't been able to get any further with this because the files from the dependency are locked in Android Studio. If anyone knows of an easy way to make local changes to the parse jar that would be helpful.

**Comment by DrBeak1** [comment]
Update - jarred the master branch and could not reproduce this. Perhaps it's fixed by a recent commit?

**Comment by scottofthescience** [comment]
I am experiencing a problem that I think is related to this.  I have one object that is saved in a background thread.  When I use save() it throws the same ACL error at runtime that you are all reporting.  I never made any modifications to the ACL here or anywhere in any part of the code in my app.  If I clear the data and uninstall the app then re-install, it works fine until I try to do a new installation on it.  This definitely slows down my development process quite a bit.

**Comment by scottofthescience** [comment]
Update - This appears to only be happening when I add a point to a User object in this one ParseObject.  I am adding pointers to Users in exactly the same way with many other ParseObjects that seem to work fine, the only difference is that in this one I manage my own threading and call Save() instead of SaveInBackground().  I'm not exactly sure that its the same issue as this, but the results is that I get the RunTimeException about ACL keys when I have never touched the ACL keys for this object.

Should I start a new issue until I am sure that it is the same as this one?

**Comment by scottofthescience** [comment]
Update - Ok, using a compiled snapshot of the current master branch Parse 1.13.2, I was able to do a temporary hack.  The problem I was having seems to be that every call I make to ParseUser.getCurrentUser() returns a ParseUser with bad ACL keys.  Even just calling ParseUser.getCurrentUser().getACL() throws the same  "RuntimeException: only ACLs can be stored in the ACL key" I was getting everywhere else, which is weird since I am asking for the ACL not setting them.

The hacky solution that I am doing now is that everywhere that I need to get a reference to the current user, I immediately set the ACLs afterward with "currentUser.setACL(new ParseACL(currentUser));".  According to Parse, this should be an entirely redundant thing to do, but since the ACL values are bad, this seems to reset them so that I can save the reference in other objects without getting a RunTimeError or a ParseException.

**Comment by DrBeak1** [comment]
Hey @scottofthescience , thanks for looking more into this. Sounds like things could be related. I haven't had time to do more exploring -- if anyone has time, one thing to do would be to add this:
```
      String className = acl.getClass().getSimpleName();
      PLog.d("ParseObject","CLASS: " + className);
```
... into the `ParseObject.java` class `getACL()` function, to see what class is actually there. That may shed light on what is happening, in that we may be able to trace back to where its being set to some other class. If no one gets to this beforehand, I may be able to do it this weekend.

**Comment by scottofthescience** [comment]
If I have time tomorrow I may gave it a go. I'll keep you posted.

**Comment by Ralphilius** [comment]
@scottofthescience Have you got a chance to take a look on this?

**Comment by scottofthescience** [comment]
@Ralphilius not yet.

**Comment by danielbenedykt** [comment]
Happens to me the same with  ParseObject.saveAllInBackground(list, new SaveCallback() {  ...  }

**Comment by scottofthescience** [comment]
@danielbenedykt, Thank you for posting, that is very helpful. My most recent guess as to what is going on is that bad ACL keys are getting sent to the method that performs the check for valid keys when saving an object.  However, I suspect that the ACL keys for the object itself are actually good, its just either the getter for the object or the ACL validation method that is buggy.

Which version of Parse are you using?

The version I used was 1.13.2, which was downloaded and compiled from the _master_ branch.  With this version, I was able do the following to work around it.

`ParseUser currentUser = ParseUser.getCurrentUser();`
  `currentUser.setACL(new ParseACL(currentUser));`
` currentUser.saveInBackground();`

What I think might be happening is that doing this sets the ACL keys properly for this instance of hte ParseObject (in this case ParseUser object) by making entirely new ACL keys for that user. 

Maybe you can try this with your `parentObject` and `childObject`?

`childObject1.setACL(new ParseACL(currentUser));`
`childObject2.setACL(new ParseACL(currentUser));`
`ParseObject.saveAllInBackground(...)`

If you do, let me know how it goes.

**Comment by danielbenedykt** [comment]
@scottofthescience thanks.
Will try that.
I am using 1.13.1 since that is the latest I could find in Maven Central.

**Comment by mkoslacz** [comment]
I have also encountered this issue using Parse SDK for Android 1.13.1. In my case there is:
- Object A 
- Object B that has reference to A and ParseUser.

The further investigation leads to the conclusion that if you use `ParseObject#saveAllInBackground()` method, or any method that uses it, for example `ParseObject#saveAll(List<T extends ParseObject>)` to save a freshly created set of objects A and B, it results in the mentioned exception. That's consistent with previous reports.

My temporary workaround is to save my objects using sync `ParseObject#save()` one by one and handle moving this operation to background on my own. You could also use `ParseObject#saveInBackground()` methods one by one and chain these using `Task#onComplete(...)`, but this approach is much more demanding and not very neat.

Waiting for a fix as it will remove lots of ugly workarounds in my code.

**Comment by fusion44** [comment]
In [ParseObject.java](https://github.com/ParsePlatform/Parse-SDK-Android/blob/master/Parse/src/main/java/com/parse/ParseObject.java) the data is held in a field called [estimatedData](https://github.com/ParsePlatform/Parse-SDK-Android/blob/master/Parse/src/main/java/com/parse/ParseObject.java#L303) as a HashMap. The value for the "ACL" key is normally a ParseACL but when saving multiple objects in background using ParseObject.saveInBackground() the value for "ACL" key sometimes is a HashMap and throws the exception. This doesn't happen when the data is saved one by one instead of batched. Possibly this happens when the batched data is prepared?

Image of the code segment: https://www.dropbox.com/s/fxsmq4a2azodkty/ParseACLError.png?dl=0

**Comment by MohamedHussienMostafa** [comment]
Guys i'm still stuck in this issue , did any one fix it ?

**Comment by matthewkrueger** [comment]
I'm also experiencing this issue and have been for several months.  We're using the latest Parse for Android (1.13.1) and bolts (1.4.0).

It seems to me that it happens when I am reinstalling my app and logging in as a new/different user from the previous installation.  Also, I switch between a development and production backend as I'm testing things, and when I switch between the two, I frequently see this when "saveAll()" is called on a list of objects.  

This makes the app crash.  When the app is restarted, the saveAll() works fine, and objects are persisted as expected.  Please advise how we can fix this, thanks!

**Comment by DrBeak1** [comment]
phew - finally looking at this again. seems like @scottofthescience workaround works! however, if you want to add any other ACL properties you're out of luck (i.e., `acl.setRoleWriteAccess('admin')`). Seems adding anything other than the current user will still throw this error.

**Comment by DrBeak1** [comment]
Been a while, but ... @scottofthescience 's workaround seems to do the trick. Thanks for digging into this, much appreciated !

**Comment by scottofthescience** [comment]
@DrBeak1 Happy to help! I wish I had the time to fix the root of the problem, but its been a busy few months.

**Comment by matthewkrueger** [comment]
@scottofthescience  can you summarize the workaround?  You posted lots of times above and there's a lot of evolving info so I'm not sure what @DrBeak1 means by "[your] workaround seems to do the trick."

Note that I'm seeing this crash in 8% of my users over the last week, according to Fabric.io...  @DrBeak1 any chance this can be prioritized for a fix?  I've been experiencing it for months and it seems to be getting worse over time.

**Comment by scottofthescience** [comment]
I don't know what more I can tell you about the workaround that I didn't already explain in [https://github.com/ParsePlatform/Parse-SDK-Android/issues/499#issuecomment-250765167](this comment above).  The idea is that there is clearly something wrong with the ParseACL object getting created during the construction of an instance of a ParseObj.  The workaround creates a new ParseACL object using the constructor parameters that we are sure work ( ParseACL(currentUser()) ) and then we replace the ACL property of the problematic ParseObj with this new ParseACL. I have no idea if it works for 1.13.1, but it definitely allowed me to save the object when I was using 1.13.2 compiled from the source.  At the time, I was only having issues with saving a ParseUser object, so these other reports of having issues with saving many objects or using straight ParseObjects are probably a related problem, but not exactly the same.  Sorry I can't be more helpful.

**Comment by DrBeak1** [comment]
@matthewkrueger - I'm using a compiled `.jar` of Parse-Android-SDK 1.13.3. Unlike @scottofthescience , I was having issues saving `ParseObjects` rather than `ParseUser` object. However, I used @scottofthescience 's approach and it seems to be avoiding the noted RuntimeException, and everything saves as expected. Here is exactly what I'm doing in my code:

```
ParseUser currentUser = ParseUser.getCurrentUser();
currentUser.setACL(new ParseACL(currentUser));

ParseACL acl = new ParseACL(currentUser);
acl.setPublicReadAccess(false);
acl.setPublicWriteAccess(false);
acl.setRoleReadAccess("Administrator", true);
acl.setRoleWriteAccess("Administrator", true);

MyModelParseObject obj = new MyModelParseObject();
obj.setACL(acl);
```
Above is how I setup the `ACL` for each `ParseObject` that I have. In my case, I then send them using `ParseObject.saveAllInBackground(objects, completion{});`

**Comment by matthewkrueger** [comment]
Thanks @DrBeak1 .  I'm seeing the issue as well with ParseObjects, not just the ParseUser object.  It's quite widespread and happens when saving several different objects to Parse.  Hopefully this is fixed in an upcoming release.

**Comment by DrBeak1** [comment]
@matthewkrueger , honestly @scottofthescience and I have been looking at this issue since mid-last-year and no project owners/admins or whatever have chimed in - it's open source now. As far as I know scott is not a project owner/admin either; we're just devs using the SDK. That said, I'm not sure the project is being maintained by anyone in particular at this point. My best advice is to take a stab at fixing it yourself and push up a PR. Also, note that this specific issue has been 'closed' due to inactivity here on the thread; project owners/admins likely won't even look at this.

**Comment by Aurintas** [comment]
What I've noticed is that you only need to set ACLs for currentUser. I do it once after user log in and don't get any errors on the same session when saving objects.

I use a global variable for currentUser, but probably even this would work:
ParseUser currentUser = ParseUser.getCurrentUser();
ParseUser.getCurrentUser().setACL(new ParseACL(currentUser));

**Comment by danielsanfr** [comment]
Hello people. I created a small project that shows how to reproduce the problem: https://github.com/danielsanfr/ParseACLDecodeBug

@Jawnnypoo I created a PR containing a solution, I will be very happy if you can evaluate it.
